ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,189 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("IoOpenDeviceInterfaceRegistryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        NULL,
+                        0,
+                        &Size);
+    if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    fip = ExAllocatePoolWithTag(NonPagedPool,
+                                Size,
+                                TAG_IO);
+    if (!fip)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto cleanup;
+    }
+
+    Status = ZwQueryKey(InterfaceKey,
+                        KeyFullInformation,
+                        fip,
+                        Size,
+                        &Size);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Enumerate subkeys (i.e. the different device objects) */
+    for (i = 0; i < fip->SubKeys; i++)
+    {
+        Status = ZwEnumerateKey(InterfaceKey,
+                                i,
+                                KeyBasicInformation,
+                                NULL,
+                                0,
+                                &Size);
+        if (Status == STATUS_NO_MORE_ENTRIES)
+        {
+            break;
+        }
+        else if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+        {
+            DPRINT1("ZwEnumerateKey() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        bip = ExAllocatePoolWithTag(NonPagedPool,
+                                    Size,
+                                    TAG_IO);
+        if (!bip)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto cleanup;
+        }
+
+        Status = ZwEnumerateKey(InterfaceKey,
+                                i,
+                                KeyBasicInformation,
+                                bip,
+                                Size,
+                                &Size);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("ZwEnumerateKey() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        /* Check bip->Name */
+        RtlInitUnicodeString(&bipName, NULL);
+        RtlInitUnicodeString(&InLinkName, NULL);
+
+        bipName.Length = wcslen(bip->Name);
+        bipName.MaximumLength = wcslen(bip->Name);
+        bipName.Buffer = bip->Name;
+
+        AliasKeyName.Length = 0;
+        AliasKeyName.MaximumLength += bipName.Length;
+        AliasKeyName.Buffer = ExAllocatePoolWithTag(NonPagedPool,
```

Comment:
By analyzing where the returned Alias string has been allocated, on Windows, you can determine whether it's NonPagedPool or PagedPool that you need to use there.

Reply:
I'd be surprised if it wasn't `PagedPool`. Most high-level Io/Registry things are.

Reply:
`IoGetDeviceInterfaceAlias` implicitly allocates stuff in the memory pool by calling its private helper, `IopGetRegistryValue`. That helper allocates pool memory two times, the second time is done if the status returned by `ZwQueryValueKey` call is `STATUS_BUFFER_OVERFLOW` or `STATUS_BUFFER_TOO_SMALL`.

As in this screenshot of the first instance of pool allocation call:
![Capture](https://user-images.githubusercontent.com/34916900/110348075-8e0e8080-8031-11eb-841f-cb8f03c2df3f.PNG)

Which 0 == NonPagedPool.

Reply:
You are not looking at the correct function @GeoB99 . You should look at the one that allocates a new string.